_posts/2017-05-11-tensorflare.md

```diff
@@ -0,0 +1,150 @@
+
+To show the benefits of having such a system, we built a small experiment utilizing TensorFlow's capabilities. We first trained a model to run on US crime statistics using the method described in [TensorFlow's MNIST tutorial](https://www.tensorflow.org/get_started/mnist/beginners). We then used this model in both PySpark and TensorFlare to see what kind of results we get. In both, we run the following query:
+
+	SELECT
+		cluster,
+		sum(case when classification = 0 then 1 else 0 end) as class1, 
+		sum(case when classification = 1 then 1 else 0 end) as class2, 
+		sum(case when classification = 2 then 1 else 0 end) as class3, 
+		sum(case when classification = 3 then 1 else 0 end) as class4
+	FROM
+		(
+			SELECT 
+				cluster,
+				classifier(murder, assault, population, rape) as classification
+			FROM
+				data
+		)
+	GROUP BY cluster
+	ORDER BY cluster
+
+The classifier method mentioned above is a UDF to Spark -- a blackbox which the optimizer is unable to reason about. In our situation, this classifier is our ML model.
+
+This query allows us to build an accuracy matrix, in which 100% accuracy is represented by every index in the matrix other than the diagonal running from [0, 0] to [N - 1, N - 1] is 0. In our tests, we work on a very small sample size (200 data points) simply for fast testing. As such, we are actually able to obtain 100% accuracy, though this is not expected in larger results.
+
+First, we'll run this in PySpark to get a baseline:
+
+	>>> df = spark.sql(crimeQuery)
+	>>> time(df.show)
+	+-------+------+------+------+------+
+	|cluster|class1|class2|class3|class4|
+	+-------+------+------+------+------+
+	|      0|    43|     0|     0|     0|
+	|      1|     0|    49|     0|     0|
+	|      2|     0|     0|    54|     0|
+	|      3|     0|     0|     0|    54|
+	+-------+------+------+------+------+
+
+	took: 4542 ms
+
+Approximately 4.5 seconds! Not bad...maybe? As we've asked in other posts, how good is this result really?
+
+Running the same query in TensorFlare yields the following:
+
+	scala> val df = flare(crimeQuery)
+	scala> time(df.show)
+	+-------+------+------+------+------+
+	|cluster|class1|class2|class3|class4|
+	+-------+------+------+------+------+
+	|      0|    43|     0|     0|     0|
+	|      1|     0|    49|     0|     0|
+	|      2|     0|     0|    54|     0|
+	|      3|     0|     0|     0|    54|
+	+-------+------+------+------+------+
+
+	took: 1351 ms
+
+1.3 seconds, over a 3x speedup! As mentioned above, this is over a much smaller dataset than one would originally expect. Should we increase our dataset to 2000 data points, we find that PySpark takes almost a minute to run, whereas TensorFlare remains nearly constant!
+
```

Comment:
What does "nearly constant" mean? Same 1.3s?

Right now this part is easy to miss, so for casual readers it may look like the 3x is our best result. It would be nice if we could show some concrete numbers for the big dataset, and display them in a way that draws visual attention.